Publish codecs types in npm too

ergunsh opened the pull request:
We weren't including these types so they were becoming `any` for the `encode` options parameter.

What this PR does is:
* Update output folder of libSquoosh types to be `build/libsquoosh/types` to make it resemble its original code location
* Copy codecs types to `build/codecs` folder

So, when `build/libsquoosh/types/codecs.ts` references `../../codecs/webp/enc/webp_node_enc.js` it will have the correct typing.

ergunsh committed:
Publish codecs types in npm too

We weren't including these types so they were
becoming `any` for the `encode` options parameter.

developit commented on libsquoosh/lib/move-d-ts.js:
```diff
@@ -4,23 +4,51 @@ const path = require('path');
+      return getFilesRecursive(path.resolve(from, fileOrDirectory.name));
+    }),
+  );
+  return allFiles.flatMap((x) => x);
```
nit:
```suggestion
  return allFiles.flat();
```

ergunsh replied:
Done, thanks for the catch! :)

developit commented on libsquoosh/lib/move-d-ts.js:
```diff
@@ -4,23 +4,51 @@ const path = require('path');
+  });
+
+  const allFiles = await Promise.all(
+    filesOrDirectories.flatMap((fileOrDirectory) => {
```
nit: the iterator never returns arrays, so there's nothing to flatten.

```suggestion
    filesOrDirectories.map((fileOrDirectory) => {
```

ergunsh replied:
Makes a lot of sense, updated it

developit reviewed (approved):
LGTM with a couple nits.

jakearchibald commented on libsquoosh/lib/move-d-ts.js:
```diff
@@ -4,23 +4,51 @@ const path = require('path');
+    filesOrDirectories.flatMap((fileOrDirectory) => {
+      if (fileOrDirectory.isFile()) {
+        return path.resolve(from, fileOrDirectory.name);
+      }
+
+      return getFilesRecursive(path.resolve(from, fileOrDirectory.name));
```
```suggestion
      } else if (
        fileOrDirectory.isDirectory() &&
        fileOrDirectory.name !== 'node_modules'
      ) {
        return getFilesRecursive(path.resolve(from, fileOrDirectory.name));
      }
      return [];
```

Otherwise it bails out over some symlinks that are created as part of the codec builds. Skipping `node_modules` is just an optimisation.

ergunsh replied:
Oh makes a lot of sense, thanks for the catch; updated it! :)

jakearchibald committed:
Merge branch 'dev' into publish-codecs-types-too

ergunsh committed:
Address review comments

* Use `Array.prototype.flat` instead of `flatMap`
* Use `map` instead of `flatMap`
* Only get files from directories and skip copying `node_modules`

ergunsh committed:
Update "@types/node" version

ergunsh committed:
Reference "WebAssembly.d.ts" in missing types

ergunsh committed:
Move missing types to output types and import them

This change ensures that the types we reference in
our type declarations exists (EmscriptedWasm, WebAssembly etc.)
at the point of the compilation of the project that uses libSquoosh

ergunsh commented:
Hey folks, this time I made sure that the types work on a new project that also has `skipLibCheck: false`. If `skipLibCheck: false` we need to make sure that our declaration files don't have type errors.

The last three commits implement that (https://github.com/GoogleChromeLabs/squoosh/pull/1220/commits/0fb590b8f92039ed50b5e34bf8876c27259ba0b2, https://github.com/GoogleChromeLabs/squoosh/pull/1220/commits/aa82dd30e7abf725b7075a85b6b4edd7d3fdb555, https://github.com/GoogleChromeLabs/squoosh/pull/1220/commits/0f594341a76d273e10f8ff6d89c632e66ed2b1fa), can you take a look at them as well?

beeequeue commented:
Would love to have this merged!

ziemek99 commented:
Should be closed because of #1321.
